File: JAVA/2024-12-31/42587.java
import java.util.*;

public class Main {
    public static void main(String[] args) {
    }
}
class solution {
    public int solution(int[] priorities, int location) {
        int answer = 0;
        PriorityQueue<Integer> pq = new PriorityQueue<>(Collections.reverseOrder());

        for (int num : priorities) {
            pq.add(num);
        }
        while (!pq.isEmpty()) {
            for (int i = 0; i < priorities.length; i++) {
                if (priorities[i] == pq.peek()) {
                    pq.poll();
                    answer++;
                    if (i == location)
                        return answer;
                }
            }
        }
        return answer;
    }
}

// 처음 문제를 접근했을 때 Queue를 사용하지 않고 문제를 풀려고 했다. 하지만 순서를 보장하지 않아서 테스트케이스를 통과하지 못했다
// 그래서 문제 자체가 Queue/Stack 문제여서 Queue를 이용해 문제 해결

File: C++/2024-12-31/42587.cpp
#include <string>
#include <vector>
#include <queue>
#include <map>
using namespace std;

int solution(vector<int> priorities, int location) 
{
    int answer = 0;

	queue<pair<int, int>> processes;
	map<int, int, greater<int>> priority_count;

	for (int i = 0; i < priorities.size(); i++)
	{
		priority_count[priorities[i]]++;
		processes.push(make_pair(priorities[i], i));
	}

	while (!processes.empty())
	{
		auto process = processes.front();
		processes.pop();

		if (process.first != priority_count.begin()->first)
		{
			processes.push(process);
			continue;
		}

		if (--priority_count[process.first] == 0)
			priority_count.erase(process.first);

		answer++;

		if (process.second == location)
			break;
	}

    return answer;
}
